WIP: Added support for Monzo Bank

TomAFrench opened the pull request:
## Description:

This platform adds support for reading account/pot balances and the sum of the day's transactions from the Monzo API.

Currently WIP as I'm waiting on a PR being accepted upstream so I can use the proper PyPi package.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#6585

## Example entry for `configuration.yaml` (if applicable):
```yaml
monzo:
```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

TomAFrench committed:
Added monzo.py

TomAFrench committed:
Minimum viable product sensor working

TomAFrench committed:
Fixed broken indentation

TomAFrench committed:
Fixed trying to call nonexistent authd_client.client

TomAFrench committed:
Added support for tracking pot balances

TomAFrench committed:
Removed some printing of state

TomAFrench committed:
Removed commented include

TomAFrench committed:
Fixed trying to get pot name before _account_id is set

TomAFrench committed:
Various style violation fixes

TomAFrench committed:
Added monzotomtest to requirements_all.txt

TomAFrench committed:
Restructuring to conform to style guidelines

Removed orphaned "extra" member variable of MonzoSensor

TomAFrench committed:
Final style violation fixes

TomAFrench committed:
Changed to better layout which conforms to style guide

TomAFrench committed:
Added monzo.py to .coveragerc

TomAFrench committed:
Removed fitbit specific image from configurator window

Updated configurator text

TomAFrench committed:
Improved readability of adding sensors for pots

TomAFrench committed:
Only track open pots

TomAFrench committed:
Fixed syntax issues

MartinHjelmare commented on homeassistant/components/sensor/monzo.py:
```diff
@@ -0,0 +1,364 @@
+def request_app_setup(hass, config, add_entities, config_path,
+                      discovery_info=None):
+    """Assist user with configuring the Monzo dev application."""
+    configurator = hass.components.configurator
```
The configurator is deprecated. We want to use config entries now.

https://developers.home-assistant.io/docs/en/config_entries_index.html

Solije replied:
Ok, I'll have a look at that and update. Thanks.

TomAFrench committed:
Initial working example using config entries.

Can set up sensors (all sensors are set up. ignoring configuration.yaml)

There exists extraneous code due to experimentation

TomAFrench committed:
Fix: Can now choose which sensors to show

TomAFrench committed:
Removed unused Nest methods and references to flows

TomAFrench committed:
Split setting up Oauth callback view into separate function

Stopped explicitly passing hass into MonzoFlowHandler

TomAFrench committed:
Can now fully set up component from configuration page in HA

client_id and client_secret now saved in config entry

TomAFrench committed:
Fix code style issues

TomAFrench committed:
Removed print statements

TomAFrench committed:
No longer writes a .conf file when authenticating

TomAFrench committed:
Removed unneccesary config_entries check

TomAFrench committed:
Fix: readded needed import

TomAFrench committed:
Removed commented code to write to conf file

TomAFrench committed:
Sorted imports

TomAFrench committed:
Cleanup of unused definitions and small mistakes

fabaff commented on homeassistant/components/monzo/__init__.py:
```diff
@@ -0,0 +1,209 @@
+    refresh_token = config_entry.data['tokens']['refresh_token']
+    last_saved_at = config_entry.data['tokens']['last_saved_at']
+
+    print("Making a Monzo OAuth")
```
Stale print.

fabaff commented on homeassistant/components/monzo/__init__.py:
```diff
@@ -0,0 +1,209 @@
+        return True
+
+    conf = config[DOMAIN]
+    client_id = conf.get(CONF_CLIENT_ID, None)
```
`None` is not needed, it's the default.

fabaff commented on homeassistant/components/monzo/__init__.py:
```diff
@@ -0,0 +1,209 @@
+
+    conf = config[DOMAIN]
+    client_id = conf.get(CONF_CLIENT_ID, None)
+    client_secret = conf.get(CONF_CLIENT_SECRET, None)
```
Dito

fabaff commented on homeassistant/components/monzo/__init__.py:
```diff
@@ -0,0 +1,209 @@
+        DOMAIN, context={'source': config_entries.SOURCE_IMPORT},
+        data={
+            'client_id': client_id,
+            'client_secret': client_secret,
```
Use for both the defined constants.

TomAFrench committed:
Fixing issues brought up by houndci

fabaff commented on homeassistant/components/monzo/config_flow.py:
```diff
@@ -0,0 +1,148 @@
+from .local_auth import MonzoAuthCallbackView
+
+CONF_CLIENT_ID = 'client_id'
+CONF_CLIENT_SECRET = 'client_secret'
```
Both are use in `__init__.py`. Consider to move them to a `const.py` file.

fabaff commented on homeassistant/components/monzo/local_auth.py:
```diff
@@ -0,0 +1,99 @@
+"""Local Nest authentication."""
```
Stale docstring.

fabaff reviewed (commented):
Just some quick comments.

Solije commented on homeassistant/components/monzo/__init__.py:
```diff
@@ -0,0 +1,209 @@
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.event import async_track_time_interval
+
+from . import config_flow
```
I found that removing this import resulted in errors. I'm not sure what the link is however.

TomAFrench committed:
updated requirements_all.txt for change from sensor to component

TomAFrench committed:
Moved some definitions to const.py and ensured they are used.

Let .get() default to None rather than making it explicit.

Removed some last prints and old docstrings

TomAFrench committed:
Fixed some stale docstrings

Solije commented:
Hey @fabaff, thanks for looking over my code, those issues should be fixed now.

TomAFrench committed:
updated .coveragerc

Solije commented:
Still quite deep in WIP atm but it's now in a "usable" state. (ie. from a minimal `configuration.yaml` entry you can go through the entire authentication procedure and get sensors out the other end but it isn't pretty.)

To Do:
- [ ] Update config entry when token refreshes (tokens will refresh behind the scenes but on restarting HA you get an invalid refresh token error)
- [ ] Add services in order to move money between pots
- [ ] Add additional sensor classes
- [ ] Pull names and units of sensors from Monzo API rather than hardcoding them in some cases. 

Issues:
* Clicking the link in the integrations dialog to authenticate doesn't properly redirect the user, requiring a refresh in order to be redirected
* Entry in integrations page needs to be clicked away after authenticating with Monzo
* General messiness of code. I'm sure the config flow can be much improved.

MartinHjelmare commented:
Please make the PR as small as possible by limiting added features. More features can be added later. Smaller PRs are easier to trvioand will be merged faster.

From your todo list I suggest waiting with number 2 and 3.

TomAFrench committed:
Updated upstream package

config entry is now synchronised with MonzoOAuth2Client token

Removed ability to import token from file as unnecessary

TomAFrench committed:
Fixed updating config_entry to wait for changes to be saved to file.

Moved initialisation of Monzo client to within MonzoObject

TomAFrench committed:
Added support for multiple pots

Pot sensors automatically take the names of the pots

TomAFrench committed:
Stopped MonzoObject from refreshing token on creation

Removed print statement

TomAFrench committed:
config_flow goes to init step if can't set up proper redirect

TomAFrench committed:
Style changes

TomAFrench committed:
Added support for tracking daily spend without tracking account balance

TomAFrench committed:
Marked hass for internal use in MonzoObject

TomAFrench committed:
Fixed behaviour for multiple pots so that they report the correct balance

Monzo sensors are now assigned a unique id

Removed unused entry_id member variable from MonzoSensor

TomAFrench committed:
MonzoSensor now uses definition of name() from MonzoEntity

TomAFrench committed:
Added comment

TomAFrench committed:
Remove reference to config file

TomAFrench committed:
Updated comments

TomAFrench committed:
Added a space between arguments in MonzoObject initialisation

TomAFrench committed:
MonzoSensor now uses MonzoEntity's definition of device_state_attributes()

TomAFrench committed:
Sensors now show account/pot id as an attribute

TomAFrench committed:
Change so only 'tokens' key of config_entry.data is overwritting on refresh

TomAFrench committed:
Increased default interval between refreshes to 5 mins

TomAFrench committed:
Made update interval configurable

TomAFrench committed:
flake8 compliance

TomAFrench committed:
Ensure time of last refresh is added to config entry

TomAFrench committed:
Removed unnecessary else statement

TomAFrench committed:
Added local_auth.py and const.py to .coveragerc

TomAFrench committed:
Removed unused errors

TomAFrench committed:
Added CONF_USER_ID to const.py

TomAFrench committed:
Add Monzo to integrations list

TomAFrench committed:
Updated config_flow to allow sensors to be create with no entry in configuration.yaml

TomAFrench committed:
Added some tests to config flow

TomAFrench committed:
Added prefix to MonzoSensor.unique_id()

Commit:
Merge branch 'dev' into monzo

TomAFrench committed:
houndci

TomAFrench committed:
Fixed indent for houndci

TomAFrench committed:
Fixed location of CONF_CLIENT_ID and CONF_CLIENT_SECRET in imports

TomAFrench committed:
Removed broken requirement to allow travisci to work

TomAFrench committed:
Removed bad test

TomAFrench committed:
Added async_step_init to MonzoFlowHandler

TomAFrench committed:
Added __init__.py to tests folder

TomAFrench committed:
Fix: added self to self.async_step_init

Solije commented:
We can now fully set up the sensors from within the integrations page of HA and refreshes of the token is properly reflected in the config entry so there's no bad refresh token errors.

I still have an issue with having to refresh the page in order to be redirected properly to Monzo from the link step of the config flow, that's a more general issue with the handling of links within the HA instance though. Writing this though makes me think I should possibly just send the user directly to the Monzo authentication page, it's a bit of an ugly link but it's better than the user having to refresh manually.

Still WIP as the upstream PR not accepted yet and there's still some neatening/fixes to do on things behind the scenes.

TomAFrench committed:
Updated requirements for tests

TomAFrench committed:
Readded py_ryobi_gdo to requirements_all.txt as removal doesn't fix issue

balloob commented on homeassistant/components/monzo/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+                    return await self._set_up_redirect(user_input)
+
+        data_schema = OrderedDict()
+        data_schema[vol.Required(CONF_CLIENT_ID)] = str
```
I don't want people to configure client IDs and secrets in config flows. Config flows should be user centric. So just as we did for Nest, client ID and secret should be configured in `configuration.yaml`.

Solije replied:
No problem, will do.

balloob commented on homeassistant/components/monzo/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+        """Attempt to link with the Monzo account.
+
+        Route the user to a website to authenticate with Monzo. Depending on
+        implementation type we expect a pin or an external component to
```
If there is a pin method, use that one instead of doing an oauth callback.

Solije replied:
Unfortunately there isn't, I would have preferred it if so as well.

balloob commented on homeassistant/components/monzo/local_auth.py:
```diff
@@ -0,0 +1,95 @@
+                CONF_LAST_SAVED_AT: int(time.time())
+            }
+
+        hass.async_add_job(hass.config_entries.flow.async_init(
```
async_create_task

balloob commented on homeassistant/components/monzo/local_auth.py:
```diff
@@ -0,0 +1,95 @@
+            data={
+                CONF_CLIENT_ID: self.oauth.client_id,
+                CONF_CLIENT_SECRET: self.oauth.client_secret,
+                'tokens': tokens
```
If no result, tokens was not defined.

balloob commented on homeassistant/components/monzo/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+
+    def __init__(self):
+        """Initialize the config flow."""
+        pass
```
If you don't use it, no need to define it.

balloob commented on homeassistant/components/monzo/config_flow.py:
```diff
@@ -0,0 +1,141 @@
+        """Initialize the config flow."""
+        pass
+
+    async def async_step_init(self, user_input=None):
```
You can remove this, this was a legacy thing.

balloob commented on homeassistant/components/monzo/__init__.py:
```diff
@@ -0,0 +1,220 @@
+            data=new_data)
+
+
+class MonzoEntity(Entity):
```
I don't think that this extra class really adds any value.

balloob commented on homeassistant/components/monzo/__init__.py:
```diff
@@ -0,0 +1,220 @@
+        new_data = {**self.config_entry.data,
+                    'tokens': new_token,
+                    CONF_LAST_SAVED_AT: int(time.time())}
+        self._hass.config_entries.async_update_entry(
```
Not allowed to call async methods from a sync context

balloob commented on homeassistant/components/monzo/__init__.py:
```diff
@@ -0,0 +1,220 @@
+            expires_at=expires_at,
+            refresh_callback=self.update_config_entry)
+
+        self.client = Monzo.from_oauth_session(oauth_client)
```
Is this doing any I/O ? It's instantiated within an async context so it's not allowed to do any I/O

Solije commented:
Hey, thanks for the feedback. Sorry for the async issues, it's my first time using it. I'll go through and fix those up along with the other points.

Coursework is starting to hit so might take a little bit though.

balloob commented:
This PR seems to have gone stale. Closing it. You can open it again when you finished it.

vmbruno commented:
This is such a good idea! @TomAFrench do you have any plans to reopen this PR in the future? I'd love to help get this into the next release 💯
